### PR TITLE
[breaking] Change in beaver expected file layout

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
Not docs for now, just checking if changes are OK. Goal is to allow to write
something like `beaver dlang install -t builder -d docker/builder` and `beaver install -d
docker/runner` in travis scripts.

### CODE REVIEW

**Comment by mihails-strasuns-sociomantic** [comment]
I also wonder why currently `beaver run` binds `PWD` to `PWD` and not to some fixed path like `/project`

**Comment by leandro-lucarella-sociomantic** [comment]
> I also wonder why currently beaver run binds PWD to PWD and not to some fixed path like /project

To avoid surprises if there is any path calculation before going into the docker container. I think other CIs do the same.

File: bin/docker/build [hunk 1]
```diff
@@ -5,8 +5,9 @@
 # http://www.boost.org/LICENSE_1_0.txt)
 
 # Look for a suitable image name
-img="${BEAVER_DOCKER_IMG-$(git config hub.upstream)}"
-img="${img:-${TRAVIS_REPO_SLUG:-beaver}}"
+img="${BEAVER_DOCKER_IMG:-$(git config hub.upstream)}"
+img="${img:-$TRAVIS_REPO_SLUG}"
+img="${img:-beaver}"
```
Comment (leandro-lucarella-sociomantic) [inline bin/docker/build:10]: Can you split this into a separate PR so it can hit v0.4.0?

File: bin/beaver-install [hunk 2]
```diff
@@ -5,6 +5,25 @@
...
+        BEAVER_CONTEXT_DIR="$2"
+        shift
+        shift
+        ;;
+        *)
+        shift
+        ;;
+    esac
+done
+
```
Comment (leandro-lucarella-sociomantic) [inline bin/beaver-install:35]: This looks hacky, why not using `getopt` instead?

Comment (leandro-lucarella-sociomantic) [reply bin/beaver-install:35]: Also, it would be nice to add some help message and/or update the README.

File: bin/beaver-install [hunk 3]
```diff
@@ -24,15 +24,13 @@ do
     esac
 done
 
-d="$(dirname "$0")"
-
 if test -r "Dockerfile${DIST:+.$DIST}"
 then
     # Build the docker image using .$DIST as suffix if present
-    "$d/beaver" docker build -f "Dockerfile${DIST:+.$DIST}" "$@" .
+    beaver docker build -f "Dockerfile${DIST:+.$DIST}" "$@" .
```
Comment (leandro-lucarella-sociomantic) [inline bin/beaver-install]: Why do you want to require this (`beaver` being on the `PATH`)??? It's making something that works more generally to be more restrictive for which gain?

File: bin/beaver-install [hunk 4]
```diff
@@ -24,13 +26,21 @@ do
...
+        beaver docker gen-dockerfile \
+            -i beaver.Dockerfile \
             -o beaver.Dockerfile.generated
-    beaver docker build -f beaver.Dockerfile.generated "$@" .
-fi
+        beaver docker build -f beaver.Dockerfile.generated "$@" .
+    fi
+}
+
+(cd $BEAVER_CONTEXT_DIR && generate_and_build $@)
```
Comment (leandro-lucarella-sociomantic) [inline bin/beaver-install]: I don't like this. The idea was to keep the top-level project directory as the working directory when invoking `docker build`, and to copy the `.dockerignore` file in the "context" directory to the top-level. This way the `Dockerfile` writer can have full control over what's sent to the docker daemon on build-time, and can be anything in the project. Limiting users to files located in `docker/` have proven to be problematic in some cases in the past, I wouldn't want to lose the flexibility here.

File: bin/beaver-install [hunk 5]
```diff
@@ -36,11 +43,13 @@ generate_and_build()
         beaver docker build -f "Dockerfile${DIST:+.$DIST}" "$@" .
     else
         # Build the docker image based on a beaver.Dockerfile base file
+        set -x
         beaver docker gen-dockerfile \
             -i beaver.Dockerfile \
-            -o beaver.Dockerfile.generated
+            -o beaver.Dockerfile.generated \
+            "${dockerfile_commands[@]-}"
```
Comment (leandro-lucarella-sociomantic) [inline bin/beaver-install]: What is this `-` at the end for?

File: bin/beaver-install [hunk 6]
```diff
@@ -16,7 +17,13 @@ do
...
         ;;
         -d)
-        export BEAVER_CONTEXT_DIR="$2"
+        beaver_context_dir="$2"
+        shift
+        shift
+        ;;
+        -I)
+        dockerfile_commands+=("-I")
+        dockerfile_commands+=("$2")
```
Comment (leandro-lucarella-sociomantic) [inline bin/beaver-install:26]: I wonder if it makes sense to forward arguments to `docker gen-dockerfile` manually like this of if it wouldn't make more sense to have an option to just forward any argument to it, like `dmd`s `-L` forwards arguments to the linker. So: `-I-Idocker/builder` will forward `-Idocker/builder` to `gen-dockerfile`.

**Review by leandro-lucarella-sociomantic — CHANGES_REQUESTED** [review]
The major thing here being the `cd $context` before the build, unless I'm getting something wrong, because I don't get the use of the `.dockerignore` files in the tests then, if this is really happening.

File: bin/docker/build (hunk 1 above)
Comment (mihails-strasuns-sociomantic) [reply bin/docker/build:10]: You haven't noticed https://github.com/sociomantic-tsunami/beaver/pull/45 , have you? :grumpy:

File: bin/beaver-install (hunk 2 above)
Comment (mihails-strasuns-sociomantic) [reply bin/beaver-install:35]: Is there any simple way to split on `--` with getopt?

File: bin/beaver-install (hunk 3 above)
Comment (mihails-strasuns-sociomantic) [reply bin/beaver-install]: Because it is already effectively required (beaver doesn't work when called via `submodules/bin/beaver-something`, I tried that) and this extra indirection just makes sources harder to read.

File: bin/beaver-install (hunk 2 above)
Comment (nemanja-boric-sociomantic) [reply bin/beaver-install:35]: `getopt` will automatically stop there:

```

while getopts ":ht" opt; do
  case ${opt} in
    h )
        echo "Help"
        shift
      ;;
    t ) # process option t
        echo "T"
        shift
      ;;
    \? ) echo "Usage: cmd [-h] [-t]"
      ;;
  esac
done

echo "Remainder: $@"

```

```
nemanjaboric@labs-129:/home/nemanjaboric  git:(config-scripts*) $ ./script.sh -h -t -- nemanja
Help
T
Remainder: -- nemanja
nemanjaboric@labs-129:/home/nemanjaboric  git:(config-scripts*) $ ./script.sh -h -- -t nemanja
Help
Remainder: -- -t nemanja
```

File: bin/beaver-install (hunk 4 above)
Comment (mihails-strasuns-sociomantic) [reply bin/beaver-install]: I don't understand why does it matter. It would surely matter a lot if we used "idiomatic" version but with beaver you must have some shell script to bind stages together no matter what. Copying/linking required files into context is just one extra step in such script - and keeps amount of magic involved in the process under control.

Comment (mihails-strasuns-sociomantic) [reply bin/beaver-install]: (with proposed approach `.dockerignore` files are not really relevant anymore indeed)

File: bin/beaver-install (hunk 6 above)
Comment (mihails-strasuns-sociomantic) [reply bin/beaver-install:26]: Can work. Right now I am trying to flesh out a bit more generic arg parsing library function to be reused from different beaver command so that everything works in a uniform way.

File: bin/beaver-install (hunk 5 above)
Comment (mihails-strasuns-sociomantic) [reply bin/beaver-install]: If `dockerfile_commands` is empty, turns it into an empty string instead of an error

**Comment by mihails-strasuns-sociomantic** [comment]
> To avoid surprises if there is any path calculation before going into the docker container. I think other CIs do the same.

Can you give an example? The problem I have with current approach is that it makes impossible to refer to files in project dir by an absolute path in build scripts. And if generated image modifies `WORKDIR`, it is all effectively gone.

File: bin/beaver-install (hunk 3 above)
Comment (leandro-lucarella-sociomantic) [reply bin/beaver-install]: I sometimes use it and it works for me. Maybe you hit a bug. I can see a point on simplifying things (in a way), but I'm not completely convinced on removing this possibility yet. When using it more locally is very useful to be able to use `beaver/bin/beaver` without touching the `PATH`.

**Comment by leandro-lucarella-sociomantic** [comment]
> Can you give an example?

Jenkins.

> The problem I have with current approach is that it makes impossible to refer to files in project dir by an absolute path in build scripts.

I didn't get this, why do you have to use absolute paths at all?

File: bin/beaver-install (hunk 2 above)
Comment (mihails-strasuns-sociomantic) [reply bin/beaver-install:35]: @nemanja-boric-sociomantic awesome, thanks!

File: bin/beaver-install (hunk 4 above)
Comment (mihails-strasuns-sociomantic) [reply bin/beaver-install]: Ping @leandro-lucarella-sociomantic interested in hearing how you see this

**Comment by mihails-strasuns-sociomantic** [comment]
> Jenkins

What about it? Why won't it work with something like `docker run -v $PWD:/project -w /project make test`? I don't understand what do you mean by "path calculation before container" at all, anything inside container is not supposed to know anything at all about external paths.

> I didn't get this, why do you have to use absolute paths at all?

Because if I run container with a different working dir, scripts would break. I'd much rather always write `apt install /project/build/last/pkg/*` and know that it always works if project is mounted, for example.

**Comment by leandro-lucarella-sociomantic** [comment]
> What about it? Why won't it work with something like docker run -v $PWD:/project -w /project make test? I don't understand what do you mean by "path calculation before container" at all, anything inside container is not supposed to know anything at all about external paths.

When you do something like `agent docker` in a `Jenkinsfile`, it will map the CWD with the full path the same as in the host. I guess you can override that, but I'm just showing that the default works otherwise. Also Makd sometimes uses absolute paths, so if some pathname is generated inside docker and then needs to be used outside docker, it will fail. I'm not saying is the wisest choice for Makd to use absolute paths, but it does, and is just another example where going for different paths in the host and in the guest could lead to surprises. OTOH I don't see any reasons why not to use the same path in the host and in the guest, I don't see any disadvantages to it.

**Comment by mihails-strasuns-sociomantic** [comment]
> When you do something like agent docker in a Jenkinsfile, it will map the CWD with the full path the same as in the host

That is so weird :X With both Gitlab CI and CircleCI repo is cloned inside the image. Any kind of context dependence is very anti-docker.

I can see where it comes from but to me honestly it looks like stockpiling hacks on hacks to retrofit tool for purpose it was never intended for.

> OTOH I don't see any reasons why not to use the same path in the host and in the guest, I don't see any disadvantages to it.

Disadvantage is that you may have to know that path and it is defined by host, not by image itself. It is not a problem if you always use it as your working dir and never change that but as soon as one wants to change that, things kind of collapse.

**Comment by mihails-strasuns-sociomantic** [comment]
(I am also not sure why would you want to use `agent docker` approach in Jenkins considering we aim to use beaver internally too)

**Comment by leandro-lucarella-sociomantic** [comment]
> What about it? Why won't it work with something like docker run -v $PWD:/project -w /project make test? I don't understand what do you mean by "path calculation before container" at all, anything inside container is not supposed to know anything at all about external paths.

When you do something like `agent docker` in a `Jenkinsfile`, it will map the CWD with the full path the same as in the host. I guess you can override that, but I'm just showing that the default works otherwise. Also Makd sometimes uses absolute paths, so if some pathname is generated inside docker and then needs to be used outside docker, it will fail. I'm not saying is the wisest choice for Makd to use absolute paths, but it does, and is just another example where going for different paths in the host and in the guest could lead to surprises. OTOH I don't see any reasons why not to use the same path in the host and in the guest, I don't see any disadvantages to it.

> That is so weird :X With both Gitlab CI and CircleCI repo is cloned inside the image. Any kind of context dependence is very anti-docker.

Is not depending on context, strictly speaking. Both Jenkins and Travis "allows" running stuff outside docker, and this is why there is still a difference between the host and the guest.

I understand your annoyance, but there is absolutely no gain in changing this now, only risk of breaking stuff. So I don't see why we should go that route now, specially considering that now we are dealing with Jenkins and Travis for 99% of the projects. I could see it if we were using mainly CircleCI or GitLab, but we aren't yet :)

> you may have to know that path and it is defined by host, not by image itself

What? For CIs that take care of the image by themselves and clone inside docker, this is a non-issue, because you can't even control/change that. For CIs where you can run inside or outside docker, then you just always use `$PWD` and that's it. I really don't see any issues with how things are now.

**Comment by mihails-strasuns-sociomantic** [comment]
Closing in favor of https://github.com/sociomantic-tsunami/beaver/pull/49
